### PR TITLE
populate gossipsub scores

### PR DESCRIPTION
**Pull request by sinkingsugar**
inspired by LH scores

### CODE REVIEW

**Comment by sinkingsugar** [comment]
From https://gist.github.com/blacktemplar/5c1862cb3f0e32a1a7fb0b25e79e6e2c#file-generate-scoring-params-py
Assumes around 50k validators, but looks like it can be realistic at this point...

File: beacon_chain/nimbus_beacon_node.nim [hunk 1]
```diff
@@ -365,10 +370,34 @@ func verifyFinalization(node: BeaconNode, slot: Slot) =
 proc installAttestationSubnetHandlers(node: BeaconNode, subnets: set[uint8]) =
   var attestationSubscriptions: seq[Future[void]] = @[]
 
+  const topicParams = TopicParams(
+    topicWeight: 0.015625,
```
Comment (tersec) [inline beacon_chain/nimbus_beacon_node.nim]: A general point is that where these come from isn't documented in the code. They have an arbitrary look at times, without such description. Even if/when the answer is, that's what LH does, that would be worth noting.

Comment (sinkingsugar) [reply beacon_chain/nimbus_beacon_node.nim]: Agreed, I'll add more rationale.

Are you comfortable with those consts inside proc bodies btw? Or you'd rather see them outside on top of the file?

I kinda like the contextual position but open to any opinion.

Comment (tersec) [reply beacon_chain/nimbus_beacon_node.nim]: When it's these single-use things, I agree with you and am fine with them in the procs, as-is. I usually do the same thing, only factoring them out once they have a second user. It minimizes namespace pollution and improves error messages by not giving the compiler more dead-ends to consider potentially relevant.

File: beacon_chain/nimbus_beacon_node.nim [hunk 2]
```diff
@@ -449,15 +478,66 @@ proc getAttestationHandlers(node: BeaconNode): Future[void] =
...
+  )
+
+  static:
+    # compile time validation
+    topicParams.validateParameters().tryGet()
+
+  node.network.subscribe(getAggregateAndProofsTopic(node.forkDigest), topicParams)
 
 proc addMessageHandlers(node: BeaconNode): Future[void] =
+  const blocksTopicParams = TopicParams(
```
Comment (tersec) [inline beacon_chain/nimbus_beacon_node.nim]: The usual nbc style is to group
```
const a
const b
```
as
```
const
  a
  b
```

**Comment by sinkingsugar** [comment]
Remaining to do:
- [x] A bit more local testing with new merged unstable
- [x] Add more comments

**Comment by sinkingsugar** [comment]
@tersec just pinged you for review but feel free to add more people if you think necessary.

File: beacon_chain/nimbus_beacon_node.nim [hunk 3]
```diff
@@ -368,10 +373,15 @@ func verifyFinalization(node: BeaconNode, slot: Slot) =
     # finalization occurs every slot, to 4 slots vs scheduledSlot.
     doAssert finalizedEpoch + 4 >= epoch
 
+proc getTotalValidators(node: BeaconNode): uint64 =
+  let epochRef = getEpochRef(node.chainDag, node.chainDag.head, node.chainDag.head.slot.epoch)
```
Comment (tersec) [inline beacon_chain/nimbus_beacon_node.nim]: https://github.com/status-im/nimbus-eth2/blob/596b8c6009ce893bbafcbefc74bc3edeecf967ce/beacon_chain/block_pools/spec_cache.nim#L19-L21

File: beacon_chain/nimbus_beacon_node.nim [hunk 4]
```diff
@@ -368,10 +373,15 @@ func verifyFinalization(node: BeaconNode, slot: Slot) =
...
+  epochRef.shuffled_active_validator_indices.lenu64()
+
 proc installAttestationSubnetHandlers(node: BeaconNode, subnets: set[uint8]) =
   # https://github.com/ethereum/eth2.0-specs/blob/v1.0.0/specs/phase0/p2p-interface.md#attestations-and-aggregation
+  # nimbus won't score attestation subnets for now, we just rely on block and aggregate which are more stabe and reliable
```
Comment (tersec) [inline beacon_chain/nimbus_beacon_node.nim:378]: It looks like attester slashings, proposer slashings, and voluntary exits get `basicParams`, which is presumably not the same as not scoring them?

Comment (tersec) [reply beacon_chain/nimbus_beacon_node.nim:378]: Oh, they are the same.

File: beacon_chain/nimbus_beacon_node.nim [hunk 5]
```diff
@@ -671,11 +681,60 @@ proc getAttestationSubnetHandlers(node: BeaconNode) =
...
+      firstMessageDeliveriesWeight: 1.1471603557060206,
+      firstMessageDeliveriesDecay: 0.9928302477768374,
+      firstMessageDeliveriesCap: 34.86870846001471,
+      meshMessageDeliveriesWeight: -458.31054878249114,
+      meshMessageDeliveriesDecay: 0.9716279515771061,
+      meshMessageDeliveriesThreshold: 0.6849191409056553,
+      meshMessageDeliveriesCap: 2.054757422716966,
+      meshMessageDeliveriesActivation: chronos.seconds(384),
+      meshMessageDeliveriesWindow: chronos.seconds(2),
+      meshFailurePenaltyWeight: -458.31054878249114 ,
```
Comment (tersec) [inline beacon_chain/nimbus_beacon_node.nim:697]: This and the other prioritization of blocks over aggregates seems appropriate.

File: beacon_chain/eth2_network.nim [hunk 6]
```diff
@@ -1567,7 +1567,20 @@ proc createEth2Node*(rng: ref BrHmacDrbgContext,
         p.historyLength = 6
         p.historyGossip = 3
         p.seenTTL = 385.seconds
+        p.gossipThreshold = -4000
+        p.publishThreshold = -8000
+        p.graylistThreshold = -16000
+        p.decayInterval = 12.seconds
```
Comment (tersec) [inline beacon_chain/eth2_network.nim]: Does `GossipSubParams` have an `init` more along the lines of `Foo(d: 8, dLow: 6, ...)` available? This is really just one single object construction, which seems better not to do as a couple dozen assignment statements.

File: beacon_chain/nimbus_beacon_node.nim (hunk 3 above)
Comment (sinkingsugar) [reply beacon_chain/nimbus_beacon_node.nim]: Ah I completely forgot about this.
It's not used yet. I might just take your comment as a reminder for the possible future and remove this for now.

File: beacon_chain/nimbus_beacon_node.nim (hunk 4 above)
Comment (sinkingsugar) [reply beacon_chain/nimbus_beacon_node.nim:378]: Yep, I tested for a whole and honestly just scoring blocks and aggregates is plenty.

File: beacon_chain/eth2_network.nim (hunk 6 above)
Comment (sinkingsugar) [reply beacon_chain/eth2_network.nim]: Yeah good point.

File: beacon_chain/nimbus_beacon_node.nim (hunk 3 above)
Comment (tersec) [reply beacon_chain/nimbus_beacon_node.nim]: Yeah, that seems even more preferable.

File: beacon_chain/eth2_network.nim [hunk 7]
```diff
@@ -1563,7 +1563,7 @@ proc createEth2Node*(rng: ref BrHmacDrbgContext,
       dLow: 6,
       dHigh: 12,
       dScore: 6,
-      dOut: 6 - 1, # DLow - 1
+      dOut: 6 div 2, # DLow - 1
```
Comment (tersec) [inline beacon_chain/eth2_network.nim]: 6 div 2 isn't DLow - 1?

Comment (sinkingsugar) [reply beacon_chain/eth2_network.nim]: comment is incorrect yeah it has to be either less than dlow but not more then dlow/2

**Review by tersec — APPROVED** [review]

**Pull request merged**